### ISSUE TITLE
LLVM 6.0: wrong code on optimization (CI OSX datetime breakage)

### ISSUE THREAD

**Issue by JohanEngelen** [ldc-developers/ldc#2529]
This is the bug that is causing troubles on our Travis OSX LLVM 6.0 tester.

I've reduced it to this:
```d
// File: datum.d
struct DateTime {
    this(ubyte minute) {
        _tod = TimeOfDay(minute);
    }

    Date _date;
    TimeOfDay _tod;
}

struct Date {
    short _year = 1;
    ubyte _month;
    // one padding byte
}

struct TimeOfDay {
    this(ubyte minute) {
        _minute = minute;
    }

    ubyte _hour;
    ubyte _minute;
    ubyte _second;
    // one padding byte
}

void main() {
    auto dt = DateTime.init;
    dt._tod._minute = 30;
    auto dt2 = DateTime(30);
    assert(dt == dt2, "wutgebeurthier");
}
```

The assert succeeds with `ldc2 -run datum.d`, but fails with `ldc2 -O3 -run datum.d`.

The optimized IR looks like this:
```llvm
  store i64 32985348833281, i64* %dt2, align 8  ; Correct init value for dt2 (minute = 30)
  %6 = getelementptr inbounds %datum.DateTime, %datum.DateTime* %tmpcast, i64 0, i32 1, i32 0
  call void @llvm.memset.p0i8.i64(i8* nonnull %6, i8 0, i64 3, i32 4, i1 false) #10 ; Why is there a memset here? It overwrites the init value!
  %7 = call i32 @memcmp(i8* nonnull %4, i8* nonnull %5, i64 8)
  %8 = icmp eq i32 %7, 0
  br i1 %8, label %assertPassed, label %assertFailed
```

I believe this may be a bug in LLVM's optimizer. In that case, it is important we find out very quickly and report it to the LLVM folks. Their LLVM 6 release is coming soon.

**Comment by JohanEngelen** [comment]
Extra info probably needed for reproducing: `target triple = "x86_64-apple-macosx"`

**Comment by kinke** [comment]
We're probably making LLVM's life harder than necessary by how we're memsetting the paddings to 0. We do it by casting the aggregate pointer to an i8* and offsetting it. We should probably use a GEP into the aggregate, to the `[N x i8]` padding array fields instead.

**Comment by kinke** [comment]
@JohanEngelen: Have you taken a look at the optimized IR for Linux x64 too? I'd be surprised if this was OSX specific, but I don't have (and plan on building) a LLVM 6/7 build.

**Comment by JohanEngelen** [comment]
haven't been able to track down the issue yet :/  I'll look at the linux x64 IR too

**Comment by JohanEngelen** [comment]
https://bugs.llvm.org/show_bug.cgi?id=36129

**Comment by JohanEngelen** [comment]
This has been fixed upstream: https://reviews.llvm.org/rL323759
I'll close this issue when it is cherrypicked onto the LLVM6 branch.

**Comment by kinke** [comment]
Glad to hear they fixed it and are going to include it in v6.0.

**Comment by JohanEngelen** [comment]
Merged in LLVM 6.0.
The suboptimal code for our padding-bytes=0 IR is tracked here: https://bugs.llvm.org/show_bug.cgi?id=36211
